per-node configuration overrides

squeed opened the pull request:
<!-- Description of change -->

This adds the ability to tweak configuration on a per-node basis, as designed in #22036. Specifically, there is an InitContainer that implements the configuration resolution logic. It writes a directory that imitates ConfigMaps, so there are no code changes in the agent itself.

This adds a new type, ClilumNodeConfig, that allows for tweaking configuration flags on a fine-grained basis. For example, you could create

```yaml
apiVersion: cilium.io/v2alpha1
kind: CiliumNodeConfig
metadata:
  namespace: kube-system
  name: kube-proxy-replacement
spec:
  selector: 
    matchLabels:
      io.cilium.kube-proxy-replacement: "" 
  rawConfig:
    kube-proxy-replacement: strict
```

then add a NodeAntiAffinity to the kube-proxy daemonset. Then, one by one, label nodes with `io.cilium.kube-proxy-replacement: ""`. This will cause that node to begin using this particular configuration override.

Fixes: #22036

```release-note
Add per-node configuration overrides. There is a new Kubernetes resource type, CiliumNodeConfig, which allows for fine-grained configuration of Nodes based on label selectors.
```

(Note: This also includes a CI workaround for the extra roles; I'll fix that in cilium-cli and remove it shortly).

squeed committed:
CRDs: add CiliumNodeConfig CRD and scaffolding

This generates the CiliumNodeConfig type, a new way to set
configuration overrides on a set of nodes.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
cmd: add build-config command

The build-config command is responsible for doing configuration
resolution on the node.

By default, it retrieves ConfigMaps, CiliumConfigOverrides, and Node
objects. However, the list of sources is customizable. The intended
usage is to allow administrators to roll changes out in a controlled
manner to a cluster.

This implements the proposed "Per-node configuration overrides" feature.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
helm: wire in the config-builder

This changes the agent daemonset to use the config-builder rather than
directly reading from the ConfigMap. This adds an initContainer that
does the configuration resolution and writes to a temporary directory.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
Documentation: add additional place were CRDs must be referenced

The existing documentation missed a point where new CRDs need to be
added.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

kaworu commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -103,7 +103,7 @@ spec:
         - cilium-agent
         args:
-        - --config-dir=/tmp/cilium/config-map
+        - --config-dir=/tmp/cilium/config-final
```
nit: this change is needed because `/tmp/cilium/config-final` is the `build-config.go` default dest right? If correct, how about either a comment or explicitly `--dest` a few lines below?

squeed replied:
You know what? This change isn't needed. I'll get rid of it.

kaworu reviewed (changes requested):
Nice work @squeed !

[This CI complain](https://github.com/cilium/cilium/actions/runs/3505020248/jobs/5871077821) seems legit:

> ciliumnodeconfigs not found in /home/runner/work/cilium/cilium/install/kubernetes/cilium/Chart.yaml.
> Please update the chart to include ciliumnodeconfigs.

Other than that, Helm part of the patch LGTM. `configSources` config feels a bit complex to get right, but admittedly I don't have much context so I'll let other reviewers chim in on the build-config flags.

thorn3r commented on pkg/option/resolver/resolver.go:
```diff
@@ -0,0 +1,285 @@
+	return cm.Data, []string{fmt.Sprintf("%s:%s/%s", KindConfigMap, namespace, name)}, nil
+}
+
+// ReadConfigOverrides reads all the CiliumConfigOverride objects and returns a flattened map
```
```suggestion
// ReadNodeOverrides reads all the CiliumNodeConfig objects and returns a flattened map
```

thorn3r commented on pkg/option/resolver/resolver.go:
```diff
@@ -0,0 +1,285 @@
+	if name == "" {
+		l, err := client.CiliumV2alpha1().CiliumNodeConfigs(namespace).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return nil, nil, fmt.Errorf("could not list CiliumConfigOverride objects: %w", err)
```
```suggestion
			return nil, nil, fmt.Errorf("could not list CiliumNodeConfig objects: %w", err)
```

thorn3r commented on pkg/option/resolver/resolver.go:
```diff
@@ -0,0 +1,285 @@
+	} else {
+		o, err := client.CiliumV2alpha1().CiliumNodeConfigs(namespace).Get(ctx, name, metav1.GetOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return nil, nil, fmt.Errorf("could not retrieve CiliumConfigOverride %s/%s: %w", namespace, name, err)
```
```suggestion
			return nil, nil, fmt.Errorf("could not retrieve CiliumNodeConfig %s/%s: %w", namespace, name, err)
```

thorn3r commented on install/kubernetes/cilium/templates/cilium-preflight/clusterrole.yaml:
```diff
@@ -70,6 +70,7 @@ rules:
   - ciliumbgppeeringpolicies
   - ciliumclusterwideenvoyconfigs
   - ciliumclusterwidenetworkpolicies
+  - ciliumconfigoverrides
```
```suggestion
  - ciliumnodeconfigs
```

thorn3r commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -510,6 +510,36 @@ daemon:
+  # This is ignored if allowedConfigOverrides is set.
+  #
+  # By default, all keys may be overridden.
+  blockedConfigOverrides: ~
```
from a user-perspective, an `any` key like `blockedConfigOverrides: *` might be nice vs. setting an invalid key for `allowedConfigOverrides` to block overrides

squeed replied:
very good point; I'll add that sugar.

squeed replied:
I added "allowedConfigOverrides: none" (which isn't actually a code change), and at Joe's suggestion, also suggested disabling overrides above.

thorn3r reviewed (changes requested):
awesome feature add, it looks great! there are still some leftover references to `CiliumConfigOverride` in comments and logs it seems. i didn't add a comment for each, but maybe `git grep -i CiliumConfigOverride`

joestringer commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumnodeconfigs.yaml:
```diff
@@ -0,0 +1,112 @@
+                  Each key must be a valid config-map data field (i.e. a-z, A-Z, -,
+                  _, and .)
+                type: object
+              selector:
```
nit: `nodeSelector` would be a bit more consistent with other naming schemes we've used:
```suggestion
              nodeSelector:
```

joestringer replied:
I think the description got updated but not the field name?

squeed replied:
no, I derped the json tag. Fix incoming.

joestringer commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -510,6 +510,36 @@ daemon:
   runPath: "/var/run/cilium"
 
+  # -- Configure a custom list of possible configuration override sources
+  # The default is "config-map:cilium-config,cilium-node-config,node"
```
Do you think there are likely to be any security impacts of these defaults? It feels a bit like we'll have to make some assumptions about the ServiceAccounts/ClusterRoles that k8s application developers typically have and whether this could potentially allow unintentional reconfiguration of cilium-agent configurations on the node. For instance, if node annotations can be freely customized, then that same privilege could be used to disable network policies in Cilium.

squeed replied:
I had the same thought around node annotations a few days ago. The Gateway API folks have a good design for a security-sensitive priority-inverting overrides system, which is cool but I balked on implementing.

For now, I'll just disable node annotations by default.

joestringer commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -510,6 +510,36 @@ daemon:
+  #
+  # This takes precedence over blockedConfigOverrides.
+  #
+  # By default, all keys may be overridden. To disable overrides, set this to an invalid key.
```
Is there an example invalid key that you could suggest? It's not so immediately obvious to me how I would disable this. For that matter, I'd probably just suggest to change `configSources` if the user wanted to completely disable overrides.

squeed replied:
Someone suggested '*', but this is even easier.

joestringer commented on pkg/option/resolver/resolver_test.go:
```diff
@@ -0,0 +1,35 @@
+			t.Fatalf("Unexpected value, wanted %s got %s", v, actual)
+		}
+	}
+}
```
Do you think it makes sense to have some golden tests or similar for evaluating that this package does the right thing?

squeed replied:
very good point. I'll look in to that.

squeed replied:
This will take a bit of time to write up, since it needs its own suite. Do you mind if it is in a follow-up PR?

I absolutely think full testing of this is critical, and I won't let it slide :-).

joestringer replied:
I'll trust you, sure. :innocent: https://github.com/cilium/cilium/issues/22507

joestringer commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -673,6 +710,8 @@ spec:
       {{- end }}
       {{- end }}
       volumes:
+      - name: tmp
```
nit: Might be nice to add a comment to highlight that this is declared to share the path between the config initContainer and the main container.

joestringer commented on Documentation/helm-values.rst:
```diff
@@ -1,6 +1,9 @@
   AUTO-GENERATED. Please DO NOT edit manually.
 
+.. role:: raw-html-m2r(raw)
+   :format: html
```
Curious how this got added, is this automatic with the newer m2r? I didn't notice before. Does it render correctly?

squeed replied:
I'm not sure; I definitely didn't add this manually. Is there an easy way to preview docs?

joestringer replied:
https://docs.cilium.io/en/latest/contributing/docs/docstest/

squeed replied:
Perfect. It doesn't seem to change the rendering, as far as I can see.

joestringer reviewed (changes requested):
Others seemed to be already paying closer attention to the details in pkg/option and cilium/cmd, so I skipped over those. I looked at the rest and from a higher level. Mostly just minor nits, one security-relevant question below.

kaworu reviewed (changes requested):
@squeed [Travis-CI is unhappy](https://app.travis-ci.com/github/cilium/cilium/jobs/590054284):

> HINT: to fix this, run 'make -C Documentation update-cmdref'

squeed committed:
install/kubernetes: label all RBAC objects

This is so we can delete them as part of cleanup. Also, it's good
practice, especially for non-namespaced objects (ClusterRole /
ClusterRoleBinding).

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
.github: manually clean up RBAC artifacts

We just need this until https://github.com/cilium/cilium-cli/issues/1257
is fixed.

Ths problem is that, right now, reinstalling after "cilium uninstall" is
broken.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

thorn3r reviewed (approved):
thanks for addressing those comments. LGTM!

kaworu commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -369,6 +366,8 @@ spec:
           mountPath: /var/lib/cilium/tls/hubble
           readOnly: true
         {{- end }}
+        - name: tmp
+          mountPath: /tmp
```
is `/tmp` the best path? It seems very generic to me, if it's all the same why not something like `/etc/cilium`?

kaworu reviewed (approved):
Helm changes LGTM

nebril reviewed (approved):
Looking pretty cool!

joestringer commented on install/kubernetes/cilium/README.md:
```diff
@@ -155,6 +155,9 @@ contributors across the globe, there is almost always someone available to help.
 | customCalls.enabled | bool | `false` | Enable tail call hooks for custom eBPF programs. |
+| daemon.allowedConfigOverrides | string | `nil` | allowedConfigOverrides is a list of config-map keys that can be overridden. That is to say, if this value is set, config sources (excepting the first one) can only override keys in this list.  This takes precedence over blockedConfigOverrides.  By default, all keys may be overridden. To disable overrides, set this to "none" or change the configSo ...
+| daemon.blockedConfigOverrides | string | `nil` | blockedConfigOverrides is a list of config-map keys that may not be overridden. In other words, if any of these keys appear in a configuration source excepting the first one, they will be ignored  This is ignored if allowedConfigOverrides is set.  By default, all keys may be overridden. |
+| daemon.configSources | string | `nil` | Configure a custom list of possible configuration override sources The default is "config-map:cilium-config,cilium-node-config" Possible entries are   - "configmap:<namespace>/<name>" -- a ConfigMap with name <name>, optionally at a namespace.   - "cilium-node-config:<namespace>/<name>" --  any or all CiliumNodeConfigurations. If <name> is specified,      ...
```
Hmm, formatting didn't turn out super great here. Broadly speaking I think that the explanation for this should probably live somewhere dedicated rather than in the configmap and generated to here, because this just ends up in a table column so it's gonna be quite hard to read.

Should we have a new page for this feature from the following location to explain how this works? https://docs.cilium.io/en/latest/configuration/

joestringer replied:
Possible other options here: 
- We can leave this as-is, maybe ideally improve the tool that generates this output so that the giant empty spaces are collapsed to 1-2 spaces.
- Move the explanation to docs page, then from the values.yaml and here we remove the explanation (or link to the corresponding docs page)
- Any other ideas?

squeed replied:
I definitely intend to write up some pretty extensive documentation about configuration overrides; I'll file a follow-up issue to fix this.

squeed replied:
Cleaned up the formatting a bit, #22576 for more docs.

joestringer reviewed (approved):
Latest changes make sense.

Just a couple of remaining items which you could either address here before we merge or follow up on to tidy up in a fresh PR.

joestringer commented on pkg/option/resolver/resolver.go:
```diff
@@ -0,0 +1,285 @@
+	Name      string // The name of the ConfigMap or Node, unused for Overrides
+}
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "config-resolver")
```
nit: Just FYI, as a convention we typically define the subsystem derived from the directory path, so in this case `option-resolver`.

squeed replied:
fixed.

joestringer commented on pkg/option/resolver/resolver.go:
```diff
@@ -0,0 +1,285 @@
+}
+
+func ReadConfigSource(ctx context.Context, client client.Clientset, nodeName string, source ConfigSource) (config map[string]string, descriptions []string, err error) {
+	log.Infof("Reading configuration from %s", source.String())
```
nit: We typically try to avoid formatting variations on logging statements, because those break structured logging and log aggregation systems since the strings are dynamic. In this case it doesn't matter quite as much since the code will be used from a separate commandline utility, but typically we would instead define this explicitly such as:
```suggestion
	log.WithFields(logrus.Fields{
	    logfields.Source: source,
	}).Info("Reading option configuration")
```

(Same for all Infof(), Errorf(), etc log statements in this file)

squeed replied:
OK, I added structured logging. Good catch.

joestringer replied:
There's still some cases where there's now both structured logging and formatting versions of the log calls. I think the right way forward is we just set up a linter for this though, then we can clean up any remaining cases in the codebase at that time.

joestringer reviewed (approved):
I did one last review pass on this. There's a couple of logging aspects that are written in a style we typically try to avoid, but given that the library is only used by this separate CLI it's not such a big deal. I think the next step from a project perspective would be to formalize those linters so that we are consistent with applying those principles, rather than having folks like me point out the common style during review like this. We can already see how easy it is to miss this sort of feedback.
